src/ray/common/scheduling/scheduling_alg.cc

```diff
@@ -0,0 +1,204 @@
+#include "scheduling_alg.h"
+#include <iostream>
+#include <stdlib.h>
+#include <chrono>
+
+using namespace std;
+
+ClusterResources::ClusterResources(int64_t local_node_id, const NodeResources &node_resources) {
+  local_node_id_ = local_node_id;
+  add(local_node_id, node_resources);
+}
+
+void ClusterResources::add(int64_t node_id, const NodeResources &node_resources) {
+
+  auto it = nodes_.find(node_id);
+  if (it == nodes_.end()) {
+    // This node is new. Add it to the list.
+    nodes_.emplace(node_id, node_resources);
+  } else {
+    // This node exists. Update it's resources.
```

Comment:
The semantics here are not obvious - I would have expected that if the , we would add the provided resources rather than replacing existing ones. Would be best to split this into separate functions - one that adds new nodes and one that updates resources for existing ones. If that makes the code uglier, then renaming the function to something like `AddOrReplaceResources` and leaving a clear comment should suffice.

Reply:
I'll change the name as you suggested. I think that dividing the function into two will make the code calling it more complex. If we hear a heartbeat from a node we just call this function, as we don't care whether this is the first heartbeat we hear from the node or not.